### PR TITLE
Feature/contact communication

### PR DESCRIPTION
**Pull request by ShrawanBK**

### CODE REVIEW

File: src/views/Countries/CommunicationAndPartners/CommunicationForm.tsx [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+    CommunicationFormFields,
+} from '#types';
+
+import {
+    requiredCondition,
+} from '#utils/validation';
+
+import styles from './styles.css';
+
+const mediumsList: BasicEntity[] = [
```
Comment (AdityaKhatri) [inline src/views/Countries/CommunicationAndPartners/CommunicationForm.tsx:35]: @annshress @ShrawanBK let's pull this from the server, and maybe not use an enum as I foresee new things being added in the future.

File: src/views/Countries/CommunicationAndPartners/CommunicationForm.tsx [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+                            ...finalValues as CommunicationFormFields,
+                            contact,
+                        },
+                    },
+                });
+            }
+        }, [createCommunication, updateCommunication, contact],
+    );
+
+    //  FIXME: `value` prop on `input` should not be null
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/CommunicationForm.tsx:273]: we just don't use `null` at all

File: src/views/Countries/CommunicationAndPartners/CommunicationForm.tsx [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+                    error={error?.fields?.content}
+                    disabled={disabled}
+                />
+            </div>
+            {/* TODO Add Loader */}
+            {loading && <Loading message="loading..." />}
+            <div className={styles.formButtons}>
+                <Button
+                    type="submit"
+                    name="submit"
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/CommunicationForm.tsx:340]:
```suggestion
                    name={undefined}
```

File: src/views/Countries/CommunicationAndPartners/CommunicationForm.tsx [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+                    label="Content *"
+                    value={value.content}
+                    onChange={onValueChange}
+                    name="content"
+                    error={error?.fields?.content}
+                    disabled={disabled}
+                />
+            </div>
+            {/* TODO Add Loader */}
+            {loading && <Loading message="loading..." />}
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/CommunicationForm.tsx:336]:
```suggestion
            {loading && <Loading />}
```

File: src/views/Countries/CommunicationAndPartners/CommunicationTable.tsx [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+            data: {
+                communicationList: {
+                    __typename: 'CommunicationListType',
+                    results: newResults,
+                },
+            },
+        });
+    }, [communicationsVariables]);
+
+    const [deleteCommunication,
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/CommunicationTable.tsx:177]:
```suggestion
    const [
        deleteCommunication,
```

File: src/views/Countries/CommunicationAndPartners/ContactForm.tsx [hunk 6]
```diff
@@ -0,0 +1,540 @@
...
+        },
+    );
+
+    const {
+        data: countries,
+        loading: countriesLoading,
+        error: countriesLoadingError,
+    } = useQuery<CountriesResponseFields>(GET_COUNTRIES_LIST);
+
+    const countriesList = countries?.countryList?.results ?? [];
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/ContactForm.tsx:310]:
```suggestion
    const countriesList = countries?.countryList?.results;
```

File: src/views/Countries/CommunicationAndPartners/ContactForm.tsx [hunk 7]
```diff
@@ -0,0 +1,540 @@
...
+        loading: organizationsLoading,
+        error: organizationsLoadingError,
+    } = useQuery<OrganizationsResponseFields>(GET_ORGANIZATIONS_LIST);
+
+    const organizationsList = organizations?.organizationList?.results.map(
+        (ol) => ({
+            id: ol.id,
+            name: ol.title,
+        }),
+    ) ?? [];
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/ContactForm.tsx:323]:
```suggestion
    );
```

File: src/views/Countries/CommunicationAndPartners/ContactForm.tsx [hunk 8]
```diff
@@ -0,0 +1,540 @@
...
+    const {
+        data: organizations,
+        loading: organizationsLoading,
+        error: organizationsLoadingError,
+    } = useQuery<OrganizationsResponseFields>(GET_ORGANIZATIONS_LIST);
+
+    const organizationsList = organizations?.organizationList?.results.map(
+        (ol) => ({
+            id: ol.id,
+            name: ol.title,
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/ContactForm.tsx:321]: This should be changed on the server, IMO

File: src/views/Countries/CommunicationAndPartners/ContactForm.tsx [hunk 9]
```diff
@@ -0,0 +1,540 @@
...
+            },
+            onError: (errors) => {
+                onErrorSet({
+                    $internal: errors.message,
+                });
+            },
+        },
+    );
+
+    // TODO write editContactLoading
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/ContactForm.tsx:371]: Is this still a TODO?

File: src/views/Countries/CommunicationAndPartners/ContactForm.tsx [hunk 10]
```diff
@@ -0,0 +1,540 @@
...
+                createContact({
+                    variables: {
+                        contact: finalValues as ContactFormFields,
+                    },
+                });
+            }
+        }, [createContact, updateContact],
+    );
+
+    //  FIXME: `value` prop on `input` should not be null
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/ContactForm.tsx:396]: We shouldn't be dealing with null app-wide

File: src/views/Countries/CommunicationAndPartners/ContactForm.tsx [hunk 11]
```diff
@@ -0,0 +1,540 @@
...
+    return (
+        <form
+            className={styles.form}
+            onSubmit={createSubmitHandler(validate, onErrorSet, handleSubmit)}
+        >
+            <div className={styles.twoColumnRow}>
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/ContactForm.tsx:402]: Should include error for errors.$internal

File: src/views/Countries/CommunicationAndPartners/index.tsx [hunk 12]
```diff
@@ -0,0 +1,664 @@
...
+            return;
+        }
+
+        const results = cacheContacts?.contactList.results ?? [];
+        const updatedResults = [...results].map((res) => {
+            if (res.id === contact.id) {
+                return contact;
+            }
+            return res;
+        });
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/index.tsx:265]: Just use splice

File: src/views/Countries/CommunicationAndPartners/index.tsx [hunk 13]
```diff
@@ -0,0 +1,664 @@
...
+            const stringColumn = (colName: stringKeys) => ({
+                headerCellRenderer: TableHeaderCell,
+                headerCellRendererParams: {
+                    onSortChange: setSortState,
+                    sortable: true,
+                    sortDirection: colName === validSortState.name
+                        ? validSortState.direction
+                        : undefined,
+                },
+                cellAsHeader: true,
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/index.tsx:467]: Not everything should be cellAsHeader

File: src/views/Countries/CommunicationAndPartners/index.tsx [hunk 14]
```diff
@@ -0,0 +1,664 @@
...
+            const dateColumn = (colName: stringKeys) => ({
+                headerCellRenderer: TableHeaderCell,
+                headerCellRendererParams: {
+                    onSortChange: setSortState,
+                    sortable: false,
+                    sortDirection: colName === validSortState.name
+                        ? validSortState.direction
+                        : undefined,
+                },
+                cellAsHeader: true,
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/index.tsx:482]: Not everything should be cellAsHeader

File: src/views/Countries/CommunicationAndPartners/index.tsx [hunk 15]
```diff
@@ -0,0 +1,664 @@
...
+            const organizationColumn = (colName: string) => ({
+                headerCellRenderer: TableHeaderCell,
+                headerCellRendererParams: {
+                    onSortChange: setSortState,
+                    sortable: false,
+                    sortDirection: colName === validSortState.name
+                        ? validSortState.direction
+                        : undefined,
+                },
+                cellAsHeader: true,
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/index.tsx:498]: Not everything should be cellAsHeader

File: src/views/Countries/CommunicationAndPartners/CommunicationTable.tsx [hunk 16]
```diff
@@ -0,0 +1,327 @@
...
+            const stringColumn = (colName: stringKeys) => ({
+                headerCellRenderer: TableHeaderCell,
+                headerCellRendererParams: {
+                    onSortChange: setSortState,
+                    sortable: true,
+                    sortDirection: colName === validSortState.name
+                        ? validSortState.direction
+                        : undefined,
+                },
+                cellAsHeader: true,
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/CommunicationTable.tsx:227]: Not everything should be cellAsHeader

File: src/views/Countries/CommunicationAndPartners/CommunicationTable.tsx [hunk 17]
```diff
@@ -0,0 +1,327 @@
...
+            const mediumColumn = (colName: string) => ({
+                headerCellRenderer: TableHeaderCell,
+                headerCellRendererParams: {
+                    onSortChange: setSortState,
+                    sortable: false,
+                    sortDirection: colName === validSortState.name
+                        ? validSortState.direction
+                        : undefined,
+                },
+                cellAsHeader: true,
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/CommunicationTable.tsx:243]: Not everything should be cellAsHeader

File: src/views/Countries/CommunicationAndPartners/CommunicationForm.tsx (hunk 1 above)
Comment (ShrawanBK) [reply src/views/Countries/CommunicationAndPartners/CommunicationForm.tsx:35]: @annshress

Comment (annshress) [reply src/views/Countries/CommunicationAndPartners/CommunicationForm.tsx:35]: Lets make a different PR for this. I will create an issue.

File: src/views/Countries/CommunicationAndPartners/ContactForm.tsx (hunk 8 above)
Comment (ShrawanBK) [reply src/views/Countries/CommunicationAndPartners/ContactForm.tsx:321]: @annshress

Comment (annshress) [reply src/views/Countries/CommunicationAndPartners/ContactForm.tsx:321]: Makes sense. Okay.

File: src/views/Countries/CommunicationAndPartners/ContactForm.tsx [hunk 18]
```diff
@@ -295,6 +295,9 @@ function ContactForm(props:ContactFormProps) {
                             (c: BasicEntity) => c.id,
                         ),
                         organization: contact.organization.id,
+                        email: contact.email ?? undefined,
```
Comment (AdityaKhatri) [inline src/views/Countries/CommunicationAndPartners/ContactForm.tsx:298]: If this is to remove null values, it is not necessary as we'll be removing null values from the response.

Comment (ShrawanBK) [reply src/views/Countries/CommunicationAndPartners/ContactForm.tsx:298]: Yes that was to remove null values !!

File: src/views/Countries/CommunicationAndPartners/CommunicationForm.tsx [hunk 19]
```diff
@@ -201,11 +208,20 @@ function CommunicationForm(props:CommunicationFormProps) {
                     contact: communication.contact.id,
                     dateTime: communication.dateTime ?? undefined,
                     title: communication.title ?? undefined,
```
Comment (AdityaKhatri) [inline src/views/Countries/CommunicationAndPartners/CommunicationForm.tsx]: Remove the checks

File: src/types/index.ts [hunk 20]
```diff
@@ -41,3 +41,65 @@ export interface User {
     // FIXME: role?: 'ADMIN' | 'IT_HEAD' | 'EDITOR' | 'REVIEWER' | 'GUEST';
     role?: string;
 }
+
+export interface ContactFormFields {
```
Comment (frozenhelium) [inline src/types/index.ts:45]: GraphQL related typings can be automatically generated now, for reference see [CrisisForm](https://github.com/idmc-labs/helix-client/blob/develop/src/components/CrisisForm/index.tsx)
CC: @tnagorra

Comment (ShrawanBK) [reply src/types/index.ts:45]: I knew this, but forgot to remove them !!
Removed now :laughing:

Comment (tnagorra) [inline src/types/index.ts:45]: Why have a separate FormFields interface? Should be derived from generated types

File: src/types/index.ts [hunk 21]
```diff
@@ -41,3 +41,65 @@ export interface User {
...
+    email: string;
+    phone: string;
+}
+
+export interface OrganizationEntity {
+    id: string;
+    title: string;
+}
+
+export type Gender = 'MALE' | 'FEMALE' | 'OTHER';
```
Comment (tnagorra) [inline src/types/index.ts:64]: None of these types should be explicitly defined!

File: src/views/Countries/CommunicationAndPartners/CommunicationTable.tsx [hunk 22]
```diff
@@ -0,0 +1,242 @@
...
+    direction: TableSortDirection.asc,
+};
+
+interface CommunicationListProps {
+    className? : string;
+    onShowAddCommunicationModal: () => void;
+    contactIdForCommunication: string;
+    onSetCommunicationIdOnEdit: (id: string) => void;
+    onCommunicationDelete: (id: string) => void;
+}
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/CommunicationTable.tsx:76]: Move Props just before the Component

File: src/views/Countries/CommunicationAndPartners/CommunicationTable.tsx [hunk 23]
```diff
@@ -0,0 +1,242 @@
...
+            const stringColumn = (colName: stringKeys) => ({
+                headerCellRenderer: TableHeaderCell,
+                headerCellRendererParams: {
+                    onSortChange: setSortState,
+                    sortable: true,
+                    sortDirection: colName === validSortState.name
+                        ? validSortState.direction
+                        : undefined,
+                },
+                cellAsHeader: true,
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/CommunicationTable.tsx:143]: cellAsHeader shouldn't be true here

File: src/views/Countries/CommunicationAndPartners/CommunicationTable.tsx [hunk 24]
```diff
@@ -0,0 +1,242 @@
...
+                cellRenderer: ActionCell,
+                cellRendererParams: (_, datum) => ({
+                    id: datum.id,
+                    onDelete: onCommunicationDelete,
+                    onEdit: handleSetCommunicationIdOnEdit,
+                }),
+            };
+
+            return [
+                createColumn(stringColumn, 'id', 'ID'),
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/CommunicationTable.tsx:182]: Why have a column for id?

File: src/views/Countries/CommunicationAndPartners/ContactForm.tsx [hunk 25]
```diff
@@ -0,0 +1,521 @@
...
+    countryList {
+      results {
+        id
+        name
+      }
+    }
+  }
+`;
+
+const designations: BasicEntity[] = [
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/ContactForm.tsx:56]: If these are enumerations, you should get this from the server

File: src/views/Countries/CommunicationAndPartners/ContactForm.tsx [hunk 26]
```diff
@@ -0,0 +1,521 @@
...
+        id: 'MR',
+        name: 'Mr.',
+    },
+    {
+        id: 'MS',
+        name: 'Ms.',
+    },
+];
+
+const genders: BasicEntity[] = [
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/ContactForm.tsx:67]: If these are enumerations, you should get this from the server

File: src/views/Countries/CommunicationAndPartners/ContactForm.tsx [hunk 27]
```diff
@@ -0,0 +1,521 @@
...
+                const { contact } = response;
+
+                if (!contact) {
+                    return;
+                }
+                onValueSet({
+                    ...contact,
+                    country: contact.country?.id,
+                    countriesOfOperation: contact.countriesOfOperation.map(
+                        (c: BasicEntity) => c.id,
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/ContactForm.tsx:260]: You should not explicity define type here, it should have been automatically inferred

File: src/views/Countries/CommunicationAndPartners/ContactForm.tsx [hunk 28]
```diff
@@ -0,0 +1,521 @@
...
+        loading: organizationsLoading,
+        error: organizationsLoadingError,
+    } = useQuery<OrganizationListQuery>(GET_ORGANIZATIONS_LIST);
+
+    // FIXME: After organization updated in server
+    const orgResults = organizations?.organizationList?.results;
+    const organizationsList = orgResults ? orgResults.map(
+        (ol) => ({
+            id: ol.id,
+            name: ol.title,
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/ContactForm.tsx:287]: Isn't title renamed to name in server?

File: src/views/Countries/CommunicationAndPartners/ContactForm.tsx (hunk 26 above)
Comment (ShrawanBK) [reply src/views/Countries/CommunicationAndPartners/ContactForm.tsx:67]: @annshress need API for this !!

File: src/views/Countries/CommunicationAndPartners/ContactForm.tsx (hunk 25 above)
Comment (ShrawanBK) [reply src/views/Countries/CommunicationAndPartners/ContactForm.tsx:56]: @annshress  need API for this !!

File: src/views/Countries/CommunicationAndPartners/index.tsx [hunk 29]
```diff
@@ -0,0 +1,721 @@
...
+                    children: (
+                        <>
+                            <Button
+                                name="view"
+                                onClick={() => onShowCommunicationListModal(datum.id)}
+                                icons={<IoIosChatboxes />}
+                            />
+                            <Button
+                                name="add-communication"
+                                onClick={() => onShowAddCommunicationModal(datum.id)}
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/index.tsx:609]: Do not create new functions, instead use id passed inside the component

File: src/views/Countries/CommunicationAndPartners/index.tsx [hunk 30]
```diff
@@ -0,0 +1,721 @@
...
+                                icons={<IoIosPersonAdd />}
+                            />
+                        </>
+                    ),
+                }),
+            };
+
+            return [
+                createColumn(stringColumn, 'id', 'ID'),
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/index.tsx:618]:
```suggestion
                createColumn(stringColumn, 'id', 'ID', true),
```

Comment (tnagorra) [reply src/views/Countries/CommunicationAndPartners/index.tsx:618]: Why show id column?

File: src/views/Countries/CommunicationAndPartners/index.tsx [hunk 31]
```diff
@@ -0,0 +1,721 @@
...
+                        contactIdForCommunication={contactIdForCommunication}
+                        onSetCommunicationIdOnEdit={setCommunicationIdOnEdit}
+                        onCommunicationDelete={handleCommunicationDelete}
+                    />
+                </Modal>
+            )}
+            {shouldShowAddCommunicationModal && (
+                <Modal
+                    onClose={handleHideAddCommunicationModal}
+                    heading={<h2>{communicationIdOnEdit ? 'Edit Communication' : 'Add '}</h2>}
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/index.tsx:693]:
```suggestion
                    heading={communicationIdOnEdit ? 'Edit Communication' : 'Add commmunication'}
```

File: src/views/Countries/CommunicationAndPartners/index.tsx [hunk 32]
```diff
@@ -0,0 +1,721 @@
...
+                        onUpdateCommunicationCache={handleUpdateCommunicationCache}
+                        onHideAddCommunicationModal={handleHideAddCommunicationModal}
+                        onAddCommunicationCache={handleAddCommunicationCache}
+                    />
+                </Modal>
+            )}
+            {shouldShowAddContactModal && (
+                <Modal
+                    onClose={hideAddContactModal}
+                    heading={<h2>{contactIdOnEdit ? 'Edit Contact' : 'Add New Contact'}</h2>}
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/index.tsx:707]:
```suggestion
                    heading={contactIdOnEdit ? 'Edit Contact' : 'Add New Contact'}
```

File: src/views/Countries/CommunicationAndPartners/ContactForm.tsx (hunk 28 above)
Comment (ShrawanBK) [reply src/views/Countries/CommunicationAndPartners/ContactForm.tsx:287]: @annshress, check this out !!

Comment (annshress) [reply src/views/Countries/CommunicationAndPartners/ContactForm.tsx:287]: Not yet.

File: src/views/Countries/CommunicationAndPartners/index.tsx [hunk 33]
```diff
@@ -0,0 +1,752 @@
...
+        (comPage) => {
+            setCommunicationPage(comPage);
+        }, [],
+    );
+
+    const handleSetCommunicationPageSize = useCallback(
+        (comPageSize) => {
+            setCommunicationPageSize(comPageSize);
+        }, [],
+    );
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/index.tsx:654]: We don't need these useCallback functions

File: src/views/Countries/CommunicationAndPartners/index.tsx [hunk 34]
```diff
@@ -0,0 +1,752 @@
...
+                <>
+                    <TextInput
+                        icons={<IoIosSearch />}
+                        name="search"
+                        value={contactSearch}
+                        placeholder="Search"
+                        onChange={setContactSearch}
+                    />
+                    <Button
+                        name="add"
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/index.tsx:670]: name={undefined}

File: src/views/Countries/CommunicationAndPartners/index.tsx [hunk 35]
```diff
@@ -0,0 +1,752 @@
...
+                        validSortState={validCommunicationSortState}
+                        onSetSortState={setSortState}
+                        loading={loadingCommunications}
+                    />
+                </Modal>
+            )}
+            {shouldShowAddCommunicationModal && (
+                <Modal
+                    onClose={handleHideAddCommunicationModal}
+                    heading={communicationIdOnEdit ? 'Edit Communication' : 'Add Communication'}
```
Comment (tnagorra) [inline src/views/Countries/CommunicationAndPartners/index.tsx:724]: "Add New Communication"

**Pull request merged**